File: filly-core/src/widgets/split_panes.rs
use crate::event::{Event, KeyCode};
use crate::render::{Orientation, Rect, RenderNode, RenderTree};
use crate::widget::{EventResult, Widget};
use filly_protocol::WidgetResponse;

pub struct SplitPanesWidget {
    pub orientation: Orientation,
    pub split_position: u16,
    pub first: Box<dyn Widget>,
    pub second: Box<dyn Widget>,
    pub active_pane: usize,
    dirty: bool,
}

impl SplitPanesWidget {
    pub fn new(orientation: Orientation, first: Box<dyn Widget>, second: Box<dyn Widget>) -> Self {
        Self {
            orientation,
            split_position: 50,
            first,
            second,
            active_pane: 0,
            dirty: true,
        }
    }
}

impl Widget for SplitPanesWidget {
    fn render(&self, area: Rect) -> RenderTree {
        let half = area.width / 2;
        let first_rect = Rect::new(0, 0, half, area.height);
        let second_rect = Rect::new(half, 0, area.width - half, area.height);

        let first_tree = self.first.render(first_rect);
        let second_tree = self.second.render(second_rect);

        RenderNode::Container {
            rect: area,
            background: None,
            border: crate::render::BorderStyle::None,
            padding: crate::render::EdgeInsets::zero(),
            children: vec![
                RenderNode::Container {
                    rect: first_rect,
                    background: if self.active_pane == 0 {
                        Some("darkgray".into())
                    } else {
                        None
                    },
                    border: crate::render::BorderStyle::Single,
                    padding: crate::render::EdgeInsets::zero(),
                    children: vec![first_tree],
                    role: None,
                    label: None,
                    description: None,
                },
                RenderNode::Container {
                    rect: second_rect,
                    background: if self.active_pane == 1 {
                        Some("darkgray".into())
                    } else {
                        None
                    },
                    border: crate::render::BorderStyle::Single,
                    padding: crate::render::EdgeInsets::zero(),
                    children: vec![second_tree],
                    role: None,
                    label: None,
                    description: None,
                },
            ],
            role: None,
            label: None,
            description: None,
        }
    }

    fn handle_event(&mut self, event: &Event) -> EventResult {
        if let Event::Key(code, _) = event {
            match code {
                KeyCode::F(1) | KeyCode::Left => {
                    self.active_pane = 0;
                    self.dirty = true;
                    return EventResult::Handled;
                }
                KeyCode::F(2) | KeyCode::Right => {
                    self.active_pane = 1;
                    self.dirty = true;
                    return EventResult::Handled;
                }
                _ => {}
            }
        }

        let result = if self.active_pane == 0 {
            self.first.handle_event(event)
        } else {
            self.second.handle_event(event)
        };

        if matches!(result, EventResult::Handled) {
            self.dirty = true;
        }
        result
    }

    fn is_dirty(&self) -> bool {
        self.dirty
    }
    fn clear_dirty(&mut self) {
        self.dirty = false;
    }
}

File: filly-core/widgets/split_panes.c
#include <stdio.h>
#include "split_panes.h"
#include <stdlib.h>
#include <string.h>

typedef struct {
    Orientation orientation;
    Widget *first;
    Widget *second;
    int split_position;
    int active_pane;
    bool dirty;
} SplitPanesData;

static void sp_render(Widget *self, Rect area, RenderTree *out) {
    SplitPanesData *d = (SplitPanesData *)(self + 1);
    memset(out, 0, sizeof(*out));
    out->type = RNODE_SPLIT_PANES;
    out->rect = area;
    out->split_panes.orientation = d->orientation;
    out->split_panes.split_position = d->split_position > 0 ? d->split_position : (d->orientation == ORIENT_HORIZONTAL ? area.w / 2 : area.h / 2);
    out->split_panes.first = malloc(sizeof(RenderTree));
    out->split_panes.second = malloc(sizeof(RenderTree));
    d->first->vtable.render(d->first, area, out->split_panes.first);
    d->second->vtable.render(d->second, area, out->split_panes.second);
}

static EventResult sp_handle_event(Widget *self, Event *ev, Backend *backend) {
    SplitPanesData *d = (SplitPanesData *)(self + 1);
    if (ev->type == EVENT_KEY) {
        if (ev->code == KEY_ESC)
            return event_result_response((WidgetResponse){ .result = NULL, .cancelled = false, .error = NULL });
        if (ev->code == KEY_F1) { d->active_pane = 0; d->dirty = true; return event_result_handled(); }
        if (ev->code == KEY_F2) { d->active_pane = 1; d->dirty = true; return event_result_handled(); }
    }
    EventResult r = d->active_pane == 0 ? d->first->vtable.handle_event(d->first, ev, backend) : d->second->vtable.handle_event(d->second, ev, backend);
    if (r.type != EVENT_RESULT_UNHANDLED) d->dirty = true;
    return r;
}

static bool sp_is_dirty(Widget *self) { return ((SplitPanesData *)(self + 1))->dirty; }
static void sp_clear_dirty(Widget *self) { ((SplitPanesData *)(self + 1))->dirty = false; }
static void sp_destroy(Widget *self) {
    SplitPanesData *d = (SplitPanesData *)(self + 1);
    widget_destroy(d->first); widget_destroy(d->second);
}

Widget *split_panes_widget_new(Orientation orientation, Widget *first, Widget *second) {
    Widget *w = calloc(1, sizeof(Widget) + sizeof(SplitPanesData));
    w->vtable.render = sp_render;
    w->vtable.handle_event = sp_handle_event;
    w->vtable.is_dirty = sp_is_dirty;
    w->vtable.clear_dirty = sp_clear_dirty;
    w->vtable.destroy = sp_destroy;
    SplitPanesData *d = (SplitPanesData *)(w + 1);
    d->orientation = orientation;
    d->first = first; d->second = second;
    d->split_position = 0;
    d->active_pane = 0;
    d->dirty = true;
    return w;
}